### PR TITLE
PR CMake

### PR DESCRIPTION
**Pull request by antoinefalisse**
In this PR, we modified the CMake files needed to build SimTKcommon with adolc

@chrisdembia 

Note: @chrisdembia, there is still one of your TODO in adolcTarget/CMakeLists. Perhaps you were still intending to do something. We can look into that when reviewing the PR.

---
This change is [Reviewable](https://reviewable.io/reviews/simbody/simbody/588)

### CODE REVIEW

**Comment by chrisdembia** [comment]
Great! I'll try this out locally soon. Travis is complaining about tabs.

**Comment by antoinefalisse** [comment]
I replaced the tabs by spaces. Travis is still complaining though. It is strange as it does not seem to be related to something I changed?

**Comment by chrisdembia** [comment]
Ah yes that was fixed in [this PR](https://github.com/simbody/simbody/pull/583). If you merge master into your `PR_CMake` branch, the error should go away.

**Comment by antoinefalisse** [comment]
I see, that is now fixed. Thanks!

File: SimTKcommon/adolcTarget/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+if(BUILD_UNVERSIONED_LIBRARIES)
+    add_library(${ADOLC_TARGET} SHARED
+            ${SOURCE_FILES} ${SOURCE_INCLUDE_FILES}
+            ${API_ABS_INCLUDE_FILES})
+
+    add_dependencies(${ADOLC_TARGET} PlatformFiles)
+
+    # TODO
+    target_include_directories(${ADOLC_TARGET} PRIVATE "${ADOLC_INCLUDES}")
```
Comment (chrisdembia) [inline SimTKcommon/adolcTarget/CMakeLists.txt]: Thanks for pointing out this TODO. I think we should remove the `target_include_directories(...)` lines for now (there's one below as well). The `include_directories("${ADOLC_INCLUDES}")` in the root CMakeLists.

This line is the preferred way to set include directories, but requires CMake 3.0. Our minimum required CMake version is 2.8 right now. So we'll just make this change later on.

File: cmake/FindADOLC.cmake [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  $ENV{ADOLCDIR}/include
+)
+
+find_library(ADOLC_LIBRARIES adolc PATHS $ENV{ADOLCDIR}/lib)
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(ADOLC DEFAULT_MSG
+                                  ADOLC_INCLUDES ADOLC_LIBRARIES)
+
+mark_as_advanced(ADOLC_INCLUDES ADOLC_LIBRARIES)
```
Comment (chrisdembia) [inline cmake/FindADOLC.cmake]: I've made some changes In my own copy of FindADOLC.cmake; could you use the following instead?


```
# This file was copied from Eigen.
# Eigen is primarily MPL2 licensed. See COPYING.MPL2 and these links:
#   http://www.mozilla.org/MPL/2.0/
#     http://www.mozilla.org/MPL/2.0/FAQ.html
if (ADOLC_INCLUDES AND ADOLC_LIBRARIES)
    set(ADOLC_FIND_QUIETLY TRUE)
endif (ADOLC_INCLUDES AND ADOLC_LIBRARIES)

set(ADOLC_DIR $ENV{ADOLC_DIR} CACHE PATH
    "Path to ADOL-C install directory.")

find_path(ADOLC_INCLUDES
        NAMES
        adolc/adtl.h
        PATHS
        "${ADOLC_DIR}/include"
)

find_library(ADOLC_LIBRARIES adolc PATHS
        "${ADOLC_DIR}/lib"
        "${ADOLC_DIR}/lib64"
        )

include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(ADOLC DEFAULT_MSG
                                  ADOLC_INCLUDES ADOLC_LIBRARIES)

mark_as_advanced(ADOLC_INCLUDES ADOLC_LIBRARIES)

```

Comment (chrisdembia) [reply cmake/FindADOLC.cmake]: It would then be good to tell users in the README that they can specify the location of ADOLC by setting the ADOLC_DIR CMake variable.

**Comment by antoinefalisse** [comment]
I have added a few instructions to the README (for windows only). I guess we can edit the README when we have a better idea whether adolc_feature will be merged into master. I have adapted the other files based on your suggestions.

File: README.md [hunk 3]
```diff
@@ -191,6 +193,7 @@ All needed library dependencies are provided with the Simbody installation on Wi
     * `BUILD_STATIC_LIBRARIES` builds the three libraries as static libraries, whose names will end with `_static`. Off by default. You must activate either `BUILD_DYNAMIC_LIBRARIES`, `BUILD_STATIC_LIBRARIES`, or both.
     * `BUILD_TESTS_AND_EXAMPLES_STATIC` if static libraries, and tests or examples are being built, creates statically-linked tests/examples. Can take a while to build, and it is unlikely you'll use the statically-linked libraries.
     * `BUILD_TESTS_AND_EXAMPLES_SHARED` if tests or examples are being built, creates dynamically-linked tests/examples. Unless you know what you're doing, leave this one on.
+    * `BUILD_ADOLC_LIBRARIES` to allow automatic differentation. You will have to provide the path to `C:/ADOLC-folder` in `ADOLC-DIR`.
```
Comment (chrisdembia) [inline README.md]: Please change `ADOLC-DIR` to `ADOLC_DIR`.

**Comment by chrisdembia** [comment]
@sherm1 how should we manage the reviews for these PRs

**Comment by sherm1** [comment]
This seems like a good process so far, @chrisdembia -- once you've had a look and approved I'll take a final pass. Is that OK with you?

Looks like this one is waiting for me now -- may take me a few days before I can give it some attention.

**Comment by chrisdembia** [comment]
@sherm1 that sounds like a good plan. At the end of my review, I can also make a list with items requiring your input.

Yes, this one is ready for you. Should @antoinefalisse prepare his next PR for my review while this PR waits for you, or should we only do one at a time?

**Comment by sherm1** [comment]
Doing them in parallel seems like a good idea. @antoinefalisse can base a new branch off this one if there are dependencies, and then rebase after this one is merged.

**Comment by chrisdembia** [comment]
Sounds good.

**Comment by antoinefalisse** [comment]
Great, I'll prepare that. Thanks.

File: .travis.yml [hunk 4]
```diff
@@ -12,16 +12,16 @@ matrix:
   include:
     - os: linux
       compiler: clang
-      env: BTYPE=RelWithDebInfo COV=OFF DEPLOY=yes
+      env: BTYPE=Release COV=OFF DEPLOY=yes
```
Comment (sherm1) [inline .travis.yml]: Why did you change these from RelWithDebInfo to Release? (I don't have an objection, just curious -- seems unrelated to the rest of the PR.)

File: SimTKcommon/tests/CMakeLists.txt [hunk 5]
```diff
@@ -26,27 +26,45 @@ foreach(TEST_PROG ${REGR_TESTS})
...
     if(BUILD_TESTS_AND_EXAMPLES_SHARED)
-        # Link with shared library
-        add_executable(${TEST_ROOT} ${TEST_PROG})
-        set_target_properties(${TEST_ROOT}
-        PROPERTIES
-          PROJECT_LABEL "Test_Regr - ${TEST_ROOT}")
-        target_link_libraries(${TEST_ROOT} 
-                    ${TEST_SHARED_TARGET})
-        add_test(${TEST_ROOT} ${EXECUTABLE_OUTPUT_PATH}/${TEST_ROOT})
+        if (NOT ${TEST_ROOT} MATCHES "TestAutodiffADOLCCommon")
```
Comment (sherm1) [inline SimTKcommon/tests/CMakeLists.txt]: This seems like a hack for testing -- we'll need a better way to decide when to link examples to the adolc libraries. I'm OK with having this here for now, but it needs a "TODO" and an explanatory comment saying what it does and explaining the limitations and promising a to repair later. Or if I'm misinterpreting, the comment should clarify why this is right.

File: CMakeLists.txt [hunk 6]
```diff
@@ -130,6 +130,10 @@ if(NOT BUILD_DYNAMIC_LIBRARIES AND NOT BUILD_STATIC_LIBRARIES)
         "are both set to FALSE. No compilation is possible!")
 endif()
 
+# ADOL-C libraries and tests won't be built unless this is set.
```
Comment (sherm1) [inline CMakeLists.txt]: Please add a comment noting that the ADOLC libraries are always built shared rather than static (assuming that's correct).

File: .travis.yml (hunk 4 above)
Comment (chrisdembia) [reply .travis.yml]: This change was actually from a PR I created and that is now merged. I'm not sure why it showed up in this PR's diff.

Comment (sherm1) [reply .travis.yml]: Hmmm. There were several other places that looked like spurious changes in this PR, which includes a surprisingly large number of commits. Possibly it needs to be merged with or rebased on master.

Also, the GitHub reviewing tool seems really awful. We should switch to Reviewable!

**Comment by sherm1** [comment]
@antoinefalisse, can you figure out why this PR seems to includes diffs that are already in master? Maybe just do a merge or rebase in your clone and then push to this branch?

**Comment by sherm1** [comment]
+@sherm1, +@chrisdembia for review (now in Reviewable).

---

Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-KxklWTJknXQhIyg_Bcj:b4sx3gt)*

**Comment by sherm1** [comment]
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[.travis.yml, line 15 at r1](https://reviewable.io:443/reviews/simbody/simbody/588#-Kxkl3bx7zr1CaXfvQNc-r1-15:-KxklkiCBP5rkjpio1l2:b9aawip) ([raw file](https://github.com/simbody/simbody/blob/df1c4e4bdd241e267d5bacf0ba2712e145e8835a/.travis.yml#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Hmmm. There were several other places that looked like spurious changes in this PR, which includes a surprisingly large number of commits. Possibly it needs to be merged with or rebased on master.

Also, the GitHub reviewing tool seems really awful. We should switch to Reviewable!
</blockquote></details>

I enabled reviewable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588)*

**Comment by antoinefalisse** [comment]
@sherm1, @chrisdembia 

> can you figure out why this PR seems to includes diffs that are already in master?

I think the branch `feature_adolc` should be merged with `master`. There are a few changes in this PR that are already merged in `master `but not in `feature_adolc` (eg `travis.yml`, `measure.cpp`, `timing.h`). I don't think I can do that myself.

**Comment by chrisdembia** [comment]
> I think the branch feature_adolc should be merged with master. There are a few changes in this PR that are already merged in masterbut not in feature_adolc (eg travis.yml, measure.cpp, timing.h). I don't think I can do that myself.

Ah good point. I've updated `feature_adolc`.

File: CMakeLists.txt (hunk 6 above)
Comment (antoinefalisse) [reply CMakeLists.txt]: @sherm1, I have added the line below (134).

File: SimTKcommon/tests/CMakeLists.txt (hunk 5 above)
Comment (antoinefalisse) [reply SimTKcommon/tests/CMakeLists.txt]: @sherm1, I have added comments where I used `if (NOT <> MATCHES <> )`. I will try to come up with a better way in another PR (I have added TODOs).

**Comment by chrisdembia** [comment]
@antoinefalisse let us know when this is ready for review again.

**Comment by antoinefalisse** [comment]
@chrisdembia and @sherm1, this is ready for review. I had replied to the comments of @sherm1 when he indicated that changes were needed. Is there a proper way of notifying you when I am done?

**Comment by chrisdembia** [comment]
>  Is there a proper way of notifying you when I am done?

There is no GitHub button or anything. Usually, we just add a comment to the PR saying "Ready for review again." or something like that.

File: SimTKcommon/tests/CMakeLists.txt [hunk 7]
```diff
@@ -26,27 +26,54 @@ foreach(TEST_PROG ${REGR_TESTS})
...
-        add_executable(${TEST_ROOT} ${TEST_PROG})
-        set_target_properties(${TEST_ROOT}
-        PROPERTIES
-          PROJECT_LABEL "Test_Regr - ${TEST_ROOT}")
-        target_link_libraries(${TEST_ROOT} 
-                    ${TEST_SHARED_TARGET})
-        add_test(${TEST_ROOT} ${EXECUTABLE_OUTPUT_PATH}/${TEST_ROOT})
+        # TestAutodiffADOLCCommon was designed to test the ADOL-C libraries only
+        # and is therefore excluded from the tests of the shared libraries.
+        # TODO need a better way to link tests to ADOL-C libraries.
```
Comment (chrisdembia) [inline SimTKcommon/tests/CMakeLists.txt]: One option is to have separate GLOBs for ADOLC vs non-ADOLC tests. The structure of this file could be as follows:

```
file(GLOB ALL_REGR_TESTS "*.cpp")
set(REGR_TESTS ${ALL_REGR_TESTS})
file(GLOB ADOLC_SPECIFIC_TESTS "*TestAutodiffADOLCCommon*")
list(REMOVE_ITEM REGR_TESTS ${ADOLC_SPECIFIC_TESTS})

foreach(TEST_PROG ${REGR_TESTS})
    # leave this loop as it used to be.
endforeach()

if(BUILD_ADOLC_LIBRARIES)
    set(ADOLC_REGR_TESTS ${ALL_REGR_TESTS})
    file(GLOB ADOLC_TESTS_TO_EXCLUDE "*TestXml*")
    list(REMOVE_ITEM ADOLC_REGR_TESTS ${ADOLC_TESTS_TO_EXCLUDE})
    foreach(TEST_PROG ${ADOLC_REGR_TESTS})
        # create the ADOLC tests.
    endforeach()
endif()

I am indifferent as to whether this is done in this PR or in another PR later.
```

**Comment by chrisdembia** [comment]
@sherm1 I added one comment, but this PR looks good to me.

**Comment by sherm1** [comment]
OK, thanks @chrisdembia (and @antoinefalisse!). It's on my reviewing queue.

**Comment by antoinefalisse** [comment]
@chrisdembia, based your last comment, I would then also modify the main CMakeLists since `BUILD_ADOLC_LIBRARIES` would no longer be in `if(BUILD_TESTS_AND_EXAMPLES_SHARED)`:

```
set(BUILD_TESTS_ADOLC ON CACHE BOOL
    "Control building of ADOL-C test programs.")
mark_as_advanced(BUILD_TESTS_ADOLC)
```

```
set(BUILD_TESTING ON CACHE BOOL
    "Control building of Simbody test programs.
    To actually build tests, at least one
    of BUILD_TESTS_AND_EXAMPLES_STATIC, 
    BUILD_TESTS_AND_EXAMPLES_SHARED, or BUILD_TESTS_ADOLC must be ON.")
```

```
if(BUILD_TESTING AND NOT (BUILD_TESTS_AND_EXAMPLES_STATIC OR
        BUILD_TESTS_AND_EXAMPLES_SHARED OR BUILD_TESTS_ADOLC))
    message(SEND_ERROR "No tests would be built, despite BUILD_EXAMPLES"
        "being on, because BUILD_TESTS_AND_EXAMPLES_STATIC and "
        "BUILD_TESTS_AND_EXAMPLES_SHARED and BUILD_TESTS_ADOLC are off.")
endif()
```
so that I can have:

```
if(BUILD_ADOLC_LIBRARIES AND BUILD_TESTS_ADOLC)
    set(ADOLC_REGR_TESTS ${ALL_REGR_TESTS})
    # The use of xml is not yet supported with the ADOL-C libraries and
    # TestXml is therefore excluded from the tests of the ADOL-C libraries.            
    file(GLOB ADOLC_TESTS_TO_EXCLUDE "*TestXml*")
    list(REMOVE_ITEM ADOLC_REGR_TESTS ${ADOLC_TESTS_TO_EXCLUDE})
    foreach(TEST_PROG ${ADOLC_REGR_TESTS})
        get_filename_component(TEST_ROOT ${TEST_PROG} NAME_WE)        
        # Link with adolc library
        set(TEST_ADOLC ${TEST_ROOT}ADOLC)
        add_executable(${TEST_ADOLC} ${TEST_PROG})
        set_target_properties(${TEST_ADOLC}
        PROPERTIES
        PROJECT_LABEL "Test_Regr - ${TEST_ADOLC}")
        target_link_libraries(${TEST_ADOLC}
                ${TEST_ADOLC_TARGET})
        add_test(${TEST_ADOLC} ${EXECUTABLE_OUTPUT_PATH}/${TEST_ADOLC})        
    endforeach()
endif()
```

It seems to give what I want so if you think it is good, I can add that to this PR, otherwise that can be for later.

**Comment by chrisdembia** [comment]
Oh good point... Could you avoid the `BUILD_TESTS_ADOLC` variable and change the `BUILD_ADOLC_LIBRARIES AND BUILD_TESTS_ADOLC` condition to `BUILD_ADOLC_LIBRARIES AND BUILD_TESTS_AND_EXAMPLES_SHARED`? Or do you think being able to turn off only the ADOLC tests is useful? I'm fine with the new `BUILD_TESTS_ADOLC` option if you find it useful.

**Comment by antoinefalisse** [comment]
No I don't think it is very useful. I have followed your suggestion and added that change to the PR.

**Comment by chrisdembia** [comment]
Thanks @antoinefalisse; LGTM.

**Comment by sherm1** [comment]
@antoinefalisse :lgtm: but there are still build failures to investigate.

---

Reviewed 7 of 12 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[SimTKcommon/adolcTarget/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/simbody/simbody/588#-KzLzg7A6YR3R1nS_Ry1:-KzLzg7A6YR3R1nS_Ry2:bjj2asw) ([raw file](https://github.com/simbody/simbody/blob/d5a72c48974c2ea00599b4af9a8bc529efa40789/SimTKcommon/adolcTarget/CMakeLists.txt#L2)):*
> ```CMake
> ## This whole directory exists just so I could define this extra
> ## preprocessor value.
> ```

BTW is this comment accurate here? May have been copied over accidentally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-KzM-A2wWbidb91IY1mi:b-jf0sdq)*

**Comment by antoinefalisse** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[SimTKcommon/adolcTarget/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/simbody/simbody/588#-KzLzg7A6YR3R1nS_Ry1:-KzM4wdncqMyqbIxX5of:b1tv6b6) ([raw file](https://github.com/simbody/simbody/blob/d5a72c48974c2ea00599b4af9a8bc529efa40789/SimTKcommon/adolcTarget/CMakeLists.txt#L2)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is this comment accurate here? May have been copied over accidentally.
</blockquote></details>

@sherm1, I think it has been copied over "accidentally" from the CMakeLists of sharedTarget.

I am actually not sure whether`add_definitions(-DSimTK_SimTKCOMMON_BUILDING_SHARED_LIBRARY)` is needed here. Perhaps @chrisdembia has a better idea?

---

*[SimTKcommon/tests/CMakeLists.txt, line 31 at r2](https://reviewable.io:443/reviews/simbody/simbody/588#-Kxkl3bx7zr1CaXfvQNj-r2-31:-Ky2df52V-1S4UQUhQ54:b-28bp41) ([raw file](https://github.com/simbody/simbody/blob/70bcd5801d57c4ec3299af996c296fafe26ddf73/SimTKcommon/tests/CMakeLists.txt#L31)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

One option is to have separate GLOBs for ADOLC vs non-ADOLC tests. The structure of this file could be as follows:

```cmake
file(GLOB ALL_REGR_TESTS "*.cpp")
set(REGR_TESTS ${ALL_REGR_TESTS})
file(GLOB ADOLC_SPECIFIC_TESTS "*TestAutodiffADOLCCommon*")
list(REMOVE_ITEM REGR_TESTS ${ADOLC_SPECIFIC_TESTS})

foreach(TEST_PROG ${REGR_TESTS})
    # leave this loop as it used to be.
endforeach()

if(BUILD_ADOLC_LIBRARIES)
    set(ADOLC_REGR_TESTS ${ALL_REGR_TESTS})
    file(GLOB ADOLC_TESTS_TO_EXCLUDE "*TestXml*")
    list(REMOVE_ITEM ADOLC_REGR_TESTS ${ADOLC_TESTS_TO_EXCLUDE})
    foreach(TEST_PROG ${ADOLC_REGR_TESTS})
        # create the ADOLC tests.
    endforeach()
endif()

I am indifferent as to whether this is done in this PR or in another PR later.
```
</blockquote></details>

Good idea. I think this requires bringing the following changes in the main CMakeLists:

set(BUILD_TESTS_ADOLC ON CACHE BOOL
    "Control building of ADOL-C test programs.")
mark_as_advanced(BUILD_TESTS_ADOLC)

set(BUILD_TESTING ON CACHE BOOL
    "Control building of Simbody test programs.
    To actually build tests, at least one
    of BUILD_TESTS_AND_EXAMPLES_STATIC, 
    BUILD_TESTS_AND_EXAMPLES_SHARED, or BUILD_TESTS_ADOLC must be ON.")

if(BUILD_TESTING AND NOT (BUILD_TESTS_AND_EXAMPLES_STATIC OR
        BUILD_TESTS_AND_EXAMPLES_SHARED OR BUILD_TESTS_ADOLC))
    message(SEND_ERROR "No tests would be built, despite BUILD_EXAMPLES"
        "being on, because BUILD_TESTS_AND_EXAMPLES_STATIC and "
        "BUILD_TESTS_AND_EXAMPLES_SHARED and BUILD_TESTS_ADOLC are both off.")
endif()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588)*

**Comment by antoinefalisse** [comment]
@sherm1, the build failures are due to a test that is designed for ADOL-C only. Since I haven't committed that test yet, Travis was complaining. I have commented out two lines in the CMake file so that it is fine for now. Later on, when I will add that test, I will uncomment these two lines. Sorry about that.

---

Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-KzMDxQxX-RaXEDMyJNd:b-oq47fw)*

**Comment by chrisdembia** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[SimTKcommon/adolcTarget/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/simbody/simbody/588#-KzLzg7A6YR3R1nS_Ry1:-KzQiijQg1EFlk0pAKSv:byt7v3a) ([raw file](https://github.com/simbody/simbody/blob/d5a72c48974c2ea00599b4af9a8bc529efa40789/SimTKcommon/adolcTarget/CMakeLists.txt#L2)):*
<details><summary><i>Previously, antoinefalisse wrote…</i></summary><blockquote>

@sherm1, I think it has been copied over "accidentally" from the CMakeLists of sharedTarget.

I am actually not sure whether`add_definitions(-DSimTK_SimTKCOMMON_BUILDING_SHARED_LIBRARY)` is needed here. Perhaps @chrisdembia has a better idea?
</blockquote></details>

I think it is still required, but the comment could be improved:

```
# This directory defines a target for building SimTKcommon (as a shared library)
# with SimTK::Real defined as ADOL-C's adouble.
# We could have put this target definition in the parent directory but we want to
# define the following SHARED_LIBRARY definition, and using add_definitions() to
# do this requires that we create this adolcTarget directory.
```

---

*[SimTKcommon/tests/CMakeLists.txt, line 31 at r2](https://reviewable.io:443/reviews/simbody/simbody/588#-Kxkl3bx7zr1CaXfvQNj-r2-31:-KzQiAVKImg8Z1wLuiDD:biedtfv) ([raw file](https://github.com/simbody/simbody/blob/70bcd5801d57c4ec3299af996c296fafe26ddf73/SimTKcommon/tests/CMakeLists.txt#L31)):*
<details><summary><i>Previously, antoinefalisse wrote…</i></summary><blockquote>

Good idea. I think this requires bringing the following changes in the main CMakeLists:

set(BUILD_TESTS_ADOLC ON CACHE BOOL
    "Control building of ADOL-C test programs.")
mark_as_advanced(BUILD_TESTS_ADOLC)

set(BUILD_TESTING ON CACHE BOOL
    "Control building of Simbody test programs.
    To actually build tests, at least one
    of BUILD_TESTS_AND_EXAMPLES_STATIC, 
    BUILD_TESTS_AND_EXAMPLES_SHARED, or BUILD_TESTS_ADOLC must be ON.")

if(BUILD_TESTING AND NOT (BUILD_TESTS_AND_EXAMPLES_STATIC OR
        BUILD_TESTS_AND_EXAMPLES_SHARED OR BUILD_TESTS_ADOLC))
    message(SEND_ERROR "No tests would be built, despite BUILD_EXAMPLES"
        "being on, because BUILD_TESTS_AND_EXAMPLES_STATIC and "
        "BUILD_TESTS_AND_EXAMPLES_SHARED and BUILD_TESTS_ADOLC are both off.")
endif()
</blockquote></details>

I thought we discussed this elsewhere already. Is this a different issue? You can add the `BUILD_TESTS_ADOLC` if you'd like and that's useful for you. Otherwise, it's fine if you just rely on `BUILD_TESTS_AND_EXAMPLES_SHARED`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588)*

**Comment by antoinefalisse** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[SimTKcommon/adolcTarget/CMakeLists.txt, line 2 at r3](https://reviewable.io:443/reviews/simbody/simbody/588#-KzLzg7A6YR3R1nS_Ry1:-KzQl9VjoqG53t62PwlA:bz81zos) ([raw file](https://github.com/simbody/simbody/blob/d5a72c48974c2ea00599b4af9a8bc529efa40789/SimTKcommon/adolcTarget/CMakeLists.txt#L2)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

I think it is still required, but the comment could be improved:

```
# This directory defines a target for building SimTKcommon (as a shared library)
# with SimTK::Real defined as ADOL-C's adouble.
# We could have put this target definition in the parent directory but we want to
# define the following SHARED_LIBRARY definition, and using add_definitions() to
# do this requires that we create this adolcTarget directory.
```
</blockquote></details>

Ok, I will add that comment. Thanks.

---

*[SimTKcommon/tests/CMakeLists.txt, line 31 at r2](https://reviewable.io:443/reviews/simbody/simbody/588#-Kxkl3bx7zr1CaXfvQNj-r2-31:-KzQk_LRMlCCLWvc5T0n:b-iersfl) ([raw file](https://github.com/simbody/simbody/blob/70bcd5801d57c4ec3299af996c296fafe26ddf73/SimTKcommon/tests/CMakeLists.txt#L31)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

I thought we discussed this elsewhere already. Is this a different issue? You can add the `BUILD_TESTS_ADOLC` if you'd like and that's useful for you. Otherwise, it's fine if you just rely on `BUILD_TESTS_AND_EXAMPLES_SHARED`.
</blockquote></details>

We have discussed this already. I only rely on `BUILD_TESTS_AND_EXAMPLES_SHARED` in the end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588)*

**Comment by chrisdembia** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588)*

**Comment by antoinefalisse** [comment]
@sherm1, this PR is now ready.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-KzRDkvM44BdHmfPVBAW:bxmikqi)*

**Comment by chrisdembia** [comment]
LGTM. @sherm1, the CI failure is resolved.

---

Reviewed 5 of 12 files at r1, 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-L-cfvjGVJHzAQGwyJW1:bsjfqb6)*

**Comment by sherm1** [comment]
Thanks, :lgtm:.

---

Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/simbody/simbody/588#-:-L-dmN2eL9q_XPviqwFH:bwq6wg0)*
